spec.html

```diff
@@ -648,28 +651,45 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
       </emu-alg>
     </emu-clause>
 
-    <emu-clause id="sec-async-module-execution-fulfilled" aoid="AsyncModuleExecutionFulfilled">
-      <h1><ins>AsyncModuleExecutionFulfilled ( _module_ )</ins></h1>
+    <emu-clause id="sec-gather-async-parent-completions" aoid="GatherAsyncParentCompletions">
+      <h1><ins>GatherAsyncParentCompletions ( _module_, _execList_ )</ins></h1>
       <emu-alg>
         1. Assert: _module_.[[Status]] is ~evaluated~.
         1. If _module_.[[AsyncEvaluating]] is *false*,
           1. Assert: _module_.[[EvaluationError]] is not ~empty~.
           1. Return *undefined*.
-        1. Assert: _module_.[[EvaluationError]] is ~empty~.
-        1. Set _module_.[[AsyncEvaluating]] to *false*.
         1. For each Module _m_ of _module_.[[AsyncParentModules]], do
-          1. Decrement _m_.[[PendingAsyncDependencies]] by 1.
-          1. If _m_.[[PendingAsyncDependencies]] is 0 and _m_.[[EvaluationError]] is ~empty~, then
+          1. If _execList_ does not contain _m_ and _m_.[[CycleRoot]].[[EvaluationError]] is ~empty~, then
```

Comment:
I think I have confirmed this now: we probably need to also guard here on either "AsyncEvaluating" being true, or "PendingAsyncDependencies" is greater than 0. In my test implementation I went with the later, otherwise we may run this again on a module, putting pendingAsyncDependencies into an invalid state (-1).

Reply:
Any module in AsyncParentModules cannot be a successful execution if we know the child is not successfully executed, and this invariant should prevent the case you describe?

If we called this function on an arbitrary module that was in an arbitrary state then I can appreciate that, but if the module is definitely a module that has not completed yet, then it's parent completions certainly won't have completed yet. The only case left should be if there was an error, where we are specifically just checking and handling that case already here.

Reply:
Would it help to add a `Assert: _module_.[[AsyncEvaluating]] is equal to *true*` before we first call this function on the top _module_ in AsyncModuleExecutionFulfilled?

Reply:
Ah, yes it turns out it was my mistake. Probably the assertion would help

Reply:
Ok great, added.